src/dlls/dbgshim/dbgshim.cpp

```diff
@@ -394,13 +394,29 @@ class RuntimeStartupHelper
 
         while (numTries < 25)
         {
+            hr = EnumerateCLRs(m_processId, ppHandleArray, ppStringArray, pdwArrayLength);
+
             // EnumerateCLRs uses the OS API CreateToolhelp32Snapshot which can return ERROR_BAD_LENGTH or 
             // ERROR_PARTIAL_COPY. If we get either of those, we try wait 1/10th of a second try again (that
-            // is the recommendation of the OS API owners)
-            hr = EnumerateCLRs(m_processId, ppHandleArray, ppStringArray, pdwArrayLength);
+            // is the recommendation of the OS API owners).
             if ((hr != HRESULT_FROM_WIN32(ERROR_PARTIAL_COPY)) && (hr != HRESULT_FROM_WIN32(ERROR_BAD_LENGTH)))
             {
-                break;
+                // Just return any other error or if no handles were found (which means the coreclr module wasn't found yet).
+                if (FAILED(hr) || *ppHandleArray == NULL || *pdwArrayLength <= 0)
+                {
+                    break;
+                }
+                // If EnumerateCLRs succeeded but any of the handles are null, then sleep and retry also. This fixes a race
+                // condition where dbgshim catches the coreclr module just being loaded but before g_hContinueStartupEvent 
+                // has been initialized.
+                for (int i = 0; i < (int)*pdwArrayLength; i++)
+                {
+                    HANDLE h = (*ppHandleArray)[i];
+                    if (h != NULL && h != INVALID_HANDLE_VALUE)
+                    {
+                        break;
+                    }
+                }
```

Comment:
If we don't find anything, would it make more sense to wait for the startup event to get set? (ex:

```
        CloseCLREnumeration(*ppHandleArrayOut, *ppStringArrayOut, *pdwArrayLengthOut);
        *ppHandleArrayOut = NULL;
        *ppStringArrayOut = NULL;
        *pdwArrayLengthOut = 0;

        return S_OK;
```

If we do decide to retry, we need to free the memory before doing so or we will leak.